### PR TITLE
Add ES6 Module support

### PR DESCRIPTION
**Pull request by oligriffiths**
This PR adds ESM support to the Brocfile.js
You can now use `import/export` etc in your Brocfile and in plugins (though we will need to document this so they're backwards compatible).

### CODE REVIEW

**Comment by stefanpenner** [comment]
Awesome


I wonder if we should stay closer to node, and  only support this new syntax for Brocfile.mjs files and leave Brocfile.js in classic cjs mode.

Cc @jdalton

**Comment by jdalton** [comment]
👋 @stefanpenner!

Allowing `.js` for ES modules is great! Node's experimental ES module support can, and will, change. Likely as soon as April. There will be support for ES module files other than `.mjs`, so a way for `.js` files to be treated as ES modules. However, the level of interop with CJS is still not finalized.

**Comment by oligriffiths** [comment]
> I wonder if we should stay closer to node, and only support this new syntax for Brocfile.mjs files and leave Brocfile.js in classic cjs mode.

@stefanpenner if Node plans to support .js files containing ESM, then I'd say we should do the same. Is there a downside of adding ESM support for .js files?

**Comment by oligriffiths** [comment]
it appears rn with node 11.9.0, modules only work when you include an mjs file
```
➜  node node --experimental-modules index.mjs
(node:45418) ExperimentalWarning: The ESM module loader is experimental.
foo
➜  node node --experimental-modules index.js
(node:45423) ExperimentalWarning: The ESM module loader is experimental.
/private/tmp/node/index.js:1
(function (exports, require, module, __filename, __dirname) { import foo from './foo'
                                                                     ^^^

SyntaxError: Unexpected identifier
    at new Script (vm.js:84:7)
    at createScript (vm.js:264:10)
    at Proxy.runInThisContext (vm.js:312:10)
    at Module._compile (internal/modules/cjs/loader.js:694:28)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:745:10)
    at Module.load (internal/modules/cjs/loader.js:626:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:566:12)
    at Function.Module._load (internal/modules/cjs/loader.js:558:3)
    at createDynamicModule (internal/modules/esm/translators.js:73:15)
    at Object.meta.done (internal/modules/esm/create_dynamic_module.js:40:9)
```

**Comment by stefanpenner** [comment]
@jdalton should we make both js and mid work via esm today, or should we wait re: es in us till the cjs interop is complete?

**Comment by jdalton** [comment]
I'm a big fan of making things _"just work"_ today. It's why I created `esm` 😉

**Comment by oligriffiths** [comment]
Yeah I feel like having it “just work” is good. We already have “just work” in ember with using modules and not requiring an mjs file. @stefanpenner Thoughts?

**Comment by stefanpenner** [comment]
Sounds good to me. I just wanted to double check with @jdalton as I know he’s wired into this

File: lib/load_brocfile.js [hunk 1]
```diff
@@ -29,7 +30,12 @@ module.exports = function loadBrocfile(options) {
   process.chdir(baseDir);
 
   // Load brocfile
-  const brocfile = require(brocfilePath);
+  let brocfile = esmRequire(brocfilePath);
+
+  // Esm imports require an "export default" that is returned as { default: ... }
+  if (typeof brocfile === 'object' && brocfile.default) {
```
Comment (oligriffiths) [inline lib/load_brocfile.js]: @jdalton I noticed that

```js
export default “foo”;
```
when run should esm returns
```js
{
    default: “foo”
}
```
Does this make sense then?

Comment (jdalton) [reply lib/load_brocfile.js]: Yep. Remember `null` is typeof `"object"` too. There is an undocumented option tentatively called `dedefault` which will return `default` export in cases where you only export a `default` binding.
```js
const esmRequire = require('esm')(module, {
  cjs: {
    dedefault: true
  }
});
```

Comment (jdalton) [reply lib/load_brocfile.js]: Though if you're wanting to support `.mjs` too (`esm` options don't extend to `.mjs`) you might keep your existing style check. Your check could be refined to:
```js
if (typeof brocfile === 'object' && brocfile !== null) {
  const names = Object.keys(brocfile)

  if (names.length === 0 && names[0] === 'default') {
    brocfile = brocfile.default;
  }
}
```

Comment (oligriffiths) [reply lib/load_brocfile.js]: Yeah that's an option, I'd like to keep forwards compatibility with .mjs.
I'm not sure what should happen with the following:
```js
export const FOO = 'foo';
export default tree;
```
which will produce
```js
{
  "FOO": "foo",
  "default": tree
}
```
We can either error if there are additional keys or ignore them. Technically exporting a named export won't do anything as the importing side is only looking at default so we might want to just ignore them, in which case your suggested change above would miss.

**Comment by oligriffiths** [comment]
@stefanpenner sounds like we have an accord on the way forwards. LMK if this is good.

File: lib/load_brocfile.js (hunk 1 above)
Comment (rwjblue) [reply lib/load_brocfile.js]: @oligriffiths - I'd prefer to ignore them. It seems perfectly reasonable to me for folks to have other shared exports from `Brocfile.js`...

File: README.md [hunk 2]
```diff
@@ -62,15 +62,18 @@ the following folder within your project folder:
...
 
-module.exports = () => {
- return new Funnel('app', {
+export default () => funnel('app', {
   destDir: 'appkit'
  })
 }
 ```
 
+Broccoli support ESM modules via [esm](https://www.npmjs.com/package/esm). You can also use regular CommonJS `require`
```
Comment (rwjblue) [inline README.md]:
```suggestion
Broccoli supports ES modules via [esm](https://www.npmjs.com/package/esm). You can also use regular CommonJS `require`
```

File: README.md [hunk 3]
```diff
@@ -97,29 +100,33 @@ In addition to using Broccoli via the combination of `broccoli-cli` and a `Brocf
...
+const funnel = require('broccoli-funnel');
+const merge = require('broccoli-merge');
+
+const html = funnel(appRoot, {
   files: ['index.html'],
   annotation: 'Index file'
 })
 
-const js = new Funnel(appRoot, {
+const js = funnel(appRoot, {
```
Comment (rwjblue) [inline README.md]: These changes (away from `new Foo`) seem unrelated to ESM, lets do them in a different PR.

Comment (oligriffiths) [reply README.md]: sure thing

File: lib/load_brocfile.js (hunk 1 above)
Comment (oligriffiths) [reply lib/load_brocfile.js]: cool, they are ignored by default

**Comment by oligriffiths** [comment]
@rwjblue requested changes applied
